File: Python/simple_claude/simple_api.py
"""Ultra-simple Claude API wrapper that even toddlers can use."""

import time
import sys
from typing import Optional, Callable, List
from datetime import datetime

# Rich library for beautiful, color-coded terminal output
from rich.console import Console
from rich.panel import Panel
from rich.text import Text
from rich.live import Live
from rich.layout import Layout
from rich import print as rprint
from rich.rule import Rule

try:
    from ..claude_api.api import ClaudeAPI, AsyncClaudeAPI
    from ..claude_api.models import (
        APIConfig, SessionConfig, RateLimitConfig, RetryConfig, 
        StreamConfig, DebugConfig, ClaudeModelConfig, AssistantResponse, 
        UserResponse, SystemInit, ResultSummary, TextContent, ThinkingContent,
        ToolUseContent, ToolResultContent, SystemPromptConfig
    )
except ImportError:
    # Handle case when running directly or from different directory
    import sys
    import os
    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    from claude_api.api import ClaudeAPI, AsyncClaudeAPI
    from claude_api.models import (
        APIConfig, SessionConfig, RateLimitConfig, RetryConfig, 
        StreamConfig, DebugConfig, ClaudeModelConfig, AssistantResponse, 
        UserResponse, SystemInit, ResultSummary, TextContent, ThinkingContent,
        ToolUseContent, ToolResultContent, SystemPromptConfig
    )
try:
    from .models import SimpleResponse, SimpleMetrics, SimpleConfig
except ImportError:
    # Handle case when running directly
    from models import SimpleResponse, SimpleMetrics, SimpleConfig


class SimpleClaudeAPI:
    """Ultra-simple Claude API that's foolproof and always streams with metrics."""
    
    def __init__(self, config: Optional[SimpleConfig] = None):
        """Initialize the simple Claude API.
        
        Args:
            config: Optional simple configuration. Uses sensible defaults if not provided.
        """
        self.config = config or SimpleConfig()
        self.session_metrics = SimpleMetrics(
            total_cost=0.0,
            total_tokens=0,
            requests_count=0,
            average_response_time=0.0
        )
        self._response_times: List[float] = []
        
        # Configure the underlying Claude API with sensible defaults
        api_config = APIConfig(
            session=SessionConfig(
                maintain_context=True,
                max_history_size=50,
                auto_save=True
            ),
            rate_limit=RateLimitConfig(
                requests_per_minute=30,
                requests_per_hour=500,
                enabled=True
            ),
            retry=RetryConfig(
                max_retries=self.config.max_retries,
                initial_delay=1.0,
                max_delay=10.0,
                retry_on=["TimeoutError", "ConnectionError", "RuntimeError"]
            ),
            streaming=StreamConfig(
                buffer_size=65536,  # Increased buffer size for full response capture
                timeout_between_chunks=30.0,
                yield_raw=False
            ),
            debug=DebugConfig(enabled=self.config.verbose),
            claude_model_config=ClaudeModelConfig(
                model=self.config.model,
                fallback_model="haiku" if self.config.model != "haiku" else "sonnet"  # Smart fallback
            ),
            system_prompt_config=SystemPromptConfig(
                system_prompt=self.config.system_prompt
            ),
            default_timeout=self.config.timeout_seconds
        )
        
        # Handle conversation continuation
        if self.config.auto_continue:
            api_config.conversation_config.continue_last = True
        
        self.claude_api = ClaudeAPI(api_config)
        
        # Initialize rich console for color-coded output
        self.console = Console()
        
        # Welcome message
        if not self.config.verbose:
            self.console.print("\n")
            self.console.print(Panel.fit(
                "🤖 [bold green]Simple Claude API ready![/bold green] Just call .ask() to start chatting.",
                border_style="green"
            ))
            self.console.print("")
    
    def ask(self, 
            prompt: str, 
            callback: Optional[Callable[[str], None]] = None,
            show_thinking: Optional[bool] = None) -> SimpleResponse:
        """Ask Claude a question with automatic streaming and metrics.
        
        Args:
            prompt: Your question or request to Claude
            callback: Optional function to handle streaming text (gets called with each chunk)
            show_thinking: Whether to show Claude's thinking process (overrides config)
            
        Returns:
            SimpleResponse with the complete answer and metrics
            
        Example:
            >>> claude = SimpleClaudeAPI()
            >>> response = claude.ask("What is Python?")
            >>> print(f"Cost: ${response.cost}")
        """
        if not prompt or not prompt.strip():
            self.console.print("\n")
            self.console.print("[bold red]❌ Error: Please provide a non-empty prompt![/bold red]")
            self.console.print("")
            return SimpleResponse(
                text="Error: Empty prompt provided",
                cost=0.0,
                tokens_used=0,
                model="none",
                success=False,
                tool_uses=[],
                tool_results=[]
            )
        
        self.console.print("\n")
        self.console.print(Rule("[bold blue]New Claude Request[/bold blue]", style="blue"))
        self.console.print(f"[bold cyan]🤔 Asking Claude:[/bold cyan] [white]{prompt}[/white]")
        self.console.print("[yellow]💭 Thinking and responding...[/yellow]")
        self.console.print("")
        
        start_time = time.time()
        response_text_parts = []
        thinking_parts = []
        tool_uses = []
        tool_results = []
        
        # Variables to capture metrics from streaming
        cost = 0.0
        tokens = 0
        model = "unknown"
        success = True
        session_id = None
        
        # Use custom callback or default console output with rich formatting
        def stream_handler(text: str):
            response_text_parts.append(text)
            if callback:
                callback(text)
            else:
                self.console.print(f"[green]{text}[/green]", end='', highlight=False)
        
        try:
            # Stream the response
            for event in self.claude_api.execute_stream(prompt):
                if isinstance(event, AssistantResponse):
                    # Capture model name from assistant response
                    model = event.message.model
                    for content in event.message.content:
                        if isinstance(content, TextContent):
                            stream_handler(content.text)
                        elif isinstance(content, ThinkingContent):
                            thinking_parts.append(content.thinking)
                            if show_thinking or (show_thinking is None and self.config.show_thinking):
                                self.console.print("\n")
                                self.console.print(Panel(
                                    f"[italic cyan]{content.thinking}[/italic cyan]",
                                    title="[bold yellow]💭 Claude's Thinking[/bold yellow]",
                                    border_style="yellow",
                                    expand=False
                                ))
                                self.console.print("")
                        elif isinstance(content, ToolUseContent):
                            # Capture tool use
                            tool_use_dict = {
                                "id": content.id,
                                "name": content.name,
                                "input": content.input
                            }
                            tool_uses.append(tool_use_dict)
                            
                            # Display tool use in real-time
                            self.console.print("\n")
                            self.console.print(Panel(
                                f"[bold magenta]🔧 Using Tool:[/bold magenta] [white]{content.name}[/white]\n"
                                f"[dim]Input: {str(content.input)}[/dim]",
                                title="[bold magenta]Tool Use[/bold magenta]",
                                border_style="magenta",
                                expand=False
                            ))
                            self.console.print("")
                elif isinstance(event, UserResponse):
                    # Capture tool results from user messages
                    for content in event.message.content:
                        if isinstance(content, ToolResultContent):
                            tool_result_dict = {
                                "tool_use_id": content.tool_use_id,
                                "content": content.content,
                                "is_error": content.is_error
                            }
                            tool_results.append(tool_result_dict)
                            
                            # Display tool result in real-time
                            result_style = "red" if content.is_error else "green"
                            result_icon = "❌" if content.is_error else "✅"
                            
                            self.console.print("\n")
                            self.console.print(Panel(
                                f"[bold {result_style}]{result_icon} Tool Result[/bold {result_style}]\n"
                                f"[dim]{str(content.content)}[/dim]",
                                title=f"[bold {result_style}]Tool Response[/bold {result_style}]",
                                border_style=result_style,
                                expand=False
                            ))
                            self.console.print("")
                elif isinstance(event, SystemInit):
                    # Capture session ID from system initialization
                    session_id = event.session_id
                elif isinstance(event, ResultSummary):
                    # Capture metrics from streaming response
                    cost = event.total_cost_usd
                    if event.usage:
                        tokens = event.usage.input_tokens + event.usage.output_tokens
                    success = not event.is_error
            
            end_time = time.time()
            duration = end_time - start_time
            
            # Process response
            complete_text = ''.join(response_text_parts)
            thinking_text = '\n'.join(thinking_parts) if thinking_parts else None
            
            # Try to get model info from the streaming events
            # Note: If model wasn't captured from streaming, we keep "unknown"
            
            # Update session metrics
            self.session_metrics.total_cost += cost
            self.session_metrics.total_tokens += tokens
            self.session_metrics.requests_count += 1
            self._response_times.append(duration)
            self.session_metrics.average_response_time = sum(self._response_times) / len(self._response_times)
            
            # Create response object
            simple_response = SimpleResponse(
                text=complete_text,
                cost=cost,
                tokens_used=tokens,
                model=model,
                success=success,
                session_id=session_id,
                duration_seconds=duration,
                thinking=thinking_text,
                tool_uses=tool_uses,
                tool_results=tool_results
            )
            
            # Show metrics if enabled
            if self.config.show_metrics:
                self.console.print("\n")
                self.console.print(Rule("[bold green]Response Complete[/bold green]", style="green"))
                
                # Create metrics panel
                metrics_text = (
                    f"[bold green]✅ Response completed in {duration:.1f}s[/bold green]\n"
                    f"[yellow]💰 Cost:[/yellow] [white]${cost:.4f}[/white] | "
                    f"[cyan]🎯 Tokens:[/cyan] [white]{tokens}[/white] | "
                    f"[magenta]🤖 Model:[/magenta] [white]{model}[/white]\n\n"
                    f"[bold blue]📊 Session Stats:[/bold blue]\n"
                    f"[white]💵 Total Cost: ${self.session_metrics.total_cost:.4f}[/white]\n"
                    f"[white]🎯 Total Tokens: {self.session_metrics.total_tokens}[/white]\n"
                    f"[white]📨 Requests: {self.session_metrics.requests_count}[/white]\n"
                    f"[white]⏱️  Avg Response Time: {self.session_metrics.average_response_time:.1f}s[/white]"
                )
                
                self.console.print(Panel(
                    metrics_text,
                    title="[bold blue]📊 Metrics & Performance[/bold blue]",
                    border_style="blue",
                    expand=False
                ))
                self.console.print("")
            
            return simple_response
            
        except Exception as e:
            end_time = time.time()
            duration = end_time - start_time
            
            self.console.print("\n")
            self.console.print(Panel(
                f"[bold red]❌ Something went wrong:[/bold red] [white]{str(e)}[/white]\n\n"
                f"[yellow]🔄 Don't worry, this happens sometimes. Try again![/yellow]",
                title="[bold red]Error[/bold red]",
                border_style="red",
                expand=False
            ))
            self.console.print("")
            
            # Return error response
            return SimpleResponse(
                text=f"Error: {str(e)}",
                cost=0.0,
                tokens_used=0,
                model="unknown",
                success=False,
                duration_seconds=duration,
                tool_uses=[],
                tool_results=[]
            )
    
    def chat(self, prompts: List[str]) -> List[SimpleResponse]:
        """Have a multi-turn conversation with Claude.
        
        Args:
            prompts: List of prompts to ask in sequence
            
        Returns:
            List of SimpleResponse objects
        """
        responses = []
        self.console.print("\n")
        self.console.print(Panel.fit(
            f"[bold cyan]🗣️ Starting conversation with {len(prompts)} messages...[/bold cyan]",
            border_style="cyan"
        ))
        self.console.print("")
        
        for i, prompt in enumerate(prompts, 1):
            self.console.print("\n")
            self.console.print(Rule(f"[bold purple]💬 Message {i}/{len(prompts)}[/bold purple]", style="purple"))
            response = self.ask(prompt)
            responses.append(response)
            
            if not response.success:
                self.console.print(f"[bold red]❌ Stopping conversation due to error in message {i}[/bold red]")
                break
        
        self.console.print("\n")
        self.console.print(Panel.fit(
            f"[bold green]🎉 Conversation completed! {len(responses)} responses received.[/bold green]",
            border_style="green"
        ))
        self.console.print("")
        return responses
    
    def get_metrics(self) -> SimpleMetrics:
        """Get current session metrics.
        
        Returns:
            SimpleMetrics object with current statistics
        """
        return self.session_metrics
    
    def reset_metrics(self):
        """Reset session metrics to zero."""
        self.session_metrics = SimpleMetrics(
            total_cost=0.0,
            total_tokens=0,
            requests_count=0,
            average_response_time=0.0
        )
        self._response_times.clear()
        self.console.print("[bold green]📊 Session metrics reset![/bold green]")
    
    def change_model(self, model: str):
        """Change the Claude model.
        
        Args:
            model: Model name ('sonnet', 'opus', 'haiku', or full model name)
        """
        self.config.model = model
        fallback = "haiku" if model != "haiku" else "sonnet"
        self.claude_api.set_model(model, fallback)
        self.console.print(f"[bold cyan]🤖 Switched to model:[/bold cyan] [white]{model}[/white] [dim](fallback: {fallback})[/dim]")
    
    def enable_thinking(self, enabled: bool = True):
        """Enable or disable showing Claude's thinking process.
        
        Args:
            enabled: Whether to show thinking
        """
        self.config.show_thinking = enabled
        status = "enabled" if enabled else "disabled"
        self.console.print(f"[bold yellow]🧠 Thinking display {status}[/bold yellow]")
    
    def enable_metrics(self, enabled: bool = True):
        """Enable or disable metrics display.
        
        Args:
            enabled: Whether to show metrics
        """
        self.config.show_metrics = enabled
        status = "enabled" if enabled else "disabled"
        self.console.print(f"[bold blue]📊 Metrics display {status}[/bold blue]")
    
    def set_system_prompt(self, system_prompt: Optional[str] = None):
        """Set or clear the system prompt that gets appended to all requests.
        
        Args:
            system_prompt: The system prompt text to append to all requests, or None to clear
        """
        self.config.system_prompt = system_prompt
        self.claude_api.set_system_prompt(system_prompt)
        
        if system_prompt:
            self.console.print(f"[bold green]📝 System prompt set:[/bold green] [white]{system_prompt}[/white]")
        else:
            self.console.print("[bold green]📝 System prompt cleared[/bold green]")
    
    def help(self):
        """Show help information."""
        help_text = """
🤖 Simple Claude API Help

Basic Usage:
  claude = SimpleClaudeAPI()
  response = claude.ask("Your question here")

Methods:
  .ask(prompt, callback=None)     - Ask a question (with streaming)
  .chat([prompt1, prompt2, ...])  - Multi-turn conversation
  .get_metrics()                  - View session statistics
  .reset_metrics()                - Reset statistics
  .change_model(model)            - Switch models
  .enable_thinking(True/False)    - Show/hide thinking
  .enable_metrics(True/False)     - Show/hide metrics
  .set_system_prompt(prompt)      - Set system prompt for all requests
  .help()                         - Show this help

Examples:
  # Basic usage
  claude.ask("What is Python?")
  
  # With custom callback
  def my_callback(text):
      print(f"Claude: {text}", end='')
  claude.ask("Tell me a joke", callback=my_callback)
  
  # Multiple questions
  claude.chat(["Hi!", "What's the weather like?", "Thanks!"])
  
  # With system prompt
  claude.set_system_prompt("You are a helpful Python tutor.")
  claude.ask("How do I create a list?")
  
  # Initialize with system prompt
  config = SimpleConfig(system_prompt="You are a creative writer.")
  claude = SimpleClaudeAPI(config)

Models available: 'sonnet' (default), 'opus', 'haiku'
        """
        self.console.print(Panel(
            help_text,
            title="[bold blue]🤖 Simple Claude API Help[/bold blue]",
            border_style="blue",
            expand=True
        ))


# Convenience function for one-off questions
def ask_claude_simple(prompt: str, 
                     model: str = "sonnet",
                     show_thinking: bool = False,
                     system_prompt: Optional[str] = None) -> SimpleResponse:
    """Quick function to ask Claude a single question.
    
    Args:
        prompt: Your question
        model: Model to use ('sonnet', 'opus', 'haiku')
        show_thinking: Whether to show thinking process
        system_prompt: Optional system prompt to append
        
    Returns:
        SimpleResponse with answer
    """
    config = SimpleConfig(model=model, show_thinking=show_thinking, system_prompt=system_prompt)
    claude = SimpleClaudeAPI(config)
    return claude.ask(prompt)


# Even simpler function that just prints the answer
def claude_say(prompt: str):
    """Ultra-simple function - just ask and Claude will respond.
    
    Args:
        prompt: Your question
    """
    ask_claude_simple(prompt)

File: simple_api.ts
/**
 * Ultra-simple Claude API wrapper that even toddlers can use (TypeScript version)
 */

import * as util from 'util';
import { performance } from 'perf_hooks';

import {
  ClaudeAPI,
  AsyncClaudeAPI
} from '../claude_api/api';

import {
  APIConfig,
  SessionConfig,
  RateLimitConfig,
  RetryConfig,
  StreamConfig,
  DebugConfig,
  ClaudeModelConfig,
  SystemPromptConfig,
  ConversationConfig,
  AssistantResponse,
  UserResponse,
  SystemInit,
  ResultSummary,
  TextContent,
  ThinkingContent,
  ToolUseContent,
  ToolResultContent,
} from '../claude_api/models';

import {
  SimpleResponse,
  SimpleResponseImpl,
  SimpleMetrics,
  SimpleMetricsImpl,
  SimpleConfig,
  SimpleConfigImpl,
  DEFAULT_SIMPLE_CONFIG
} from './models';

// Simple console colors for Node.js (replacement for Rich library)
const colors = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  dim: '\x1b[2m',
  
  // Basic colors
  black: '\x1b[30m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  white: '\x1b[37m',
  
  // Bright colors
  brightRed: '\x1b[91m',
  brightGreen: '\x1b[92m',
  brightYellow: '\x1b[93m',
  brightBlue: '\x1b[94m',
  brightMagenta: '\x1b[95m',
  brightCyan: '\x1b[96m',
  brightWhite: '\x1b[97m'
};

// Helper functions for console output
function colorize(text: string, color: string): string {
  return `${color}${text}${colors.reset}`;
}

function printColored(text: string, color: string = ''): void {
  console.log(colorize(text, color));
}

function printPanel(content: string, title: string = '', borderColor: string = colors.blue): void {
  const width = 80;
  const border = '─'.repeat(width - 2);
  
  console.log(colorize(`┌${border}┐`, borderColor));
  if (title) {
    const titlePadding = Math.max(0, width - title.length - 4);
    const leftPad = Math.floor(titlePadding / 2);
    const rightPad = titlePadding - leftPad;
    console.log(colorize(`│ ${' '.repeat(leftPad)}${title}${' '.repeat(rightPad)} │`, borderColor));
    console.log(colorize(`├${border}┤`, borderColor));
  }
  
  const lines = content.split('\n');
  for (const line of lines) {
    const padding = Math.max(0, width - line.length - 4);
    console.log(colorize(`│ ${line}${' '.repeat(padding)} │`, borderColor));
  }
  console.log(colorize(`└${border}┘`, borderColor));
}

function printRule(text: string, color: string = colors.blue): void {
  const width = 80;
  const textLength = text.length;
  const ruleLength = Math.max(0, width - textLength - 4);
  const leftRule = '─'.repeat(Math.floor(ruleLength / 2));
  const rightRule = '─'.repeat(Math.ceil(ruleLength / 2));
  
  console.log(colorize(`${leftRule}  ${text}  ${rightRule}`, color));
}

export class SimpleClaudeAPI {
  private config: SimpleConfig;
  private session_metrics: SimpleMetrics;
  private response_times: number[] = [];
  private claude_api: ClaudeAPI;

  constructor(config?: Partial<SimpleConfig>) {
    this.config = new SimpleConfigImpl(config);
    this.session_metrics = new SimpleMetricsImpl({
      total_cost: 0.0,
      total_tokens: 0,
      requests_count: 0,
      average_response_time: 0.0
    });

    // Configure the underlying Claude API with sensible defaults
    const api_config: Partial<APIConfig> = {
      session: {
        maintain_context: true,
        max_history_size: 50,
        auto_save: true,
        persist_to_file: undefined
      },
      rate_limit: {
        requests_per_minute: 30,
        requests_per_hour: 500,
        burst_size: 10,
        enabled: true
      },
      retry: {
        max_retries: this.config.max_retries,
        initial_delay: 1.0,
        max_delay: 10.0,
        exponential_base: 2.0,
        retry_on: ["TimeoutError", "ConnectionError", "RuntimeError"]
      },
      streaming: {
        buffer_size: 65536,
        timeout_between_chunks: 30.0,
        yield_raw: false
      },
      debug: {
        enabled: this.config.verbose
      },
      claude_model_config: {
        model: this.config.model,
        fallback_model: this.config.model !== "haiku" ? "haiku" : "sonnet"
      },
      system_prompt_config: {
        system_prompt: this.config.system_prompt
      },
      conversation_config: {
        continue_last: this.config.auto_continue,
        resume_session_id: undefined
      },
      default_timeout: this.config.timeout_seconds,
      verbose: this.config.verbose
    };

    this.claude_api = new ClaudeAPI(api_config);

    // Welcome message
    if (!this.config.verbose) {
      console.log('\n');
      printPanel(
        '🤖 Simple Claude API ready! Just call .ask() to start chatting.',
        '',
        colors.green
      );
      console.log('');
    }
  }

  /**
   * Ask Claude a question with automatic streaming and metrics
   */
  async ask(
    prompt: string,
    callback?: (text: string) => void,
    show_thinking?: boolean
  ): Promise<SimpleResponse> {
    if (!prompt || !prompt.trim()) {
      console.log('\n');
      printColored('❌ Error: Please provide a non-empty prompt!', colors.red);
      console.log('');
      return new SimpleResponseImpl({
        text: "Error: Empty prompt provided",
        cost: 0.0,
        tokens_used: 0,
        model: "none",
        success: false,
        duration_seconds: 0.0,
        tool_uses: [],
        tool_results: []
      });
    }

    console.log('\n');
    printRule('New Claude Request', colors.blue);
    printColored(`🤔 Asking Claude: ${prompt}`, colors.cyan);
    printColored('💭 Thinking and responding...', colors.yellow);
    console.log('');

    const start_time = performance.now();
    const response_text_parts: string[] = [];
    const thinking_parts: string[] = [];
    const tool_uses: Record<string, any>[] = [];
    const tool_results: Record<string, any>[] = [];

    // Variables to capture metrics from streaming
    let cost = 0.0;
    let tokens = 0;
    let model = "unknown";
    let success = true;
    let session_id: string | undefined;

    // Stream handler
    const stream_handler = (text: string) => {
      response_text_parts.push(text);
      if (callback) {
        callback(text);
      } else {
        process.stdout.write(colorize(text, colors.green));
      }
    };

    try {
      // Stream the response
      for await (const event of this.claude_api.execute_stream(prompt)) {
        if (event && typeof event === 'object' && 'type' in event) {
          if (event.type === 'assistant') {
            const assistantEvent = event as AssistantResponse;
            model = assistantEvent.message.model;
            
            for (const content of assistantEvent.message.content) {
              if (content.type === 'text') {
                const textContent = content as TextContent;
                stream_handler(textContent.text);
              } else if (content.type === 'thinking') {
                const thinkingContent = content as ThinkingContent;
                thinking_parts.push(thinkingContent.thinking);
                
                if (show_thinking || (show_thinking === undefined && this.config.show_thinking)) {
                  console.log('\n');
                  printPanel(
                    thinkingContent.thinking,
                    '💭 Claude\'s Thinking',
                    colors.yellow
                  );
                  console.log('');
                }
              } else if (content.type === 'tool_use') {
                const toolUseContent = content as ToolUseContent;
                const tool_use_dict = {
                  id: toolUseContent.id,
                  name: toolUseContent.name,
                  input: toolUseContent.input
                };
                tool_uses.push(tool_use_dict);

                console.log('\n');
                printPanel(
                  `🔧 Using Tool: ${toolUseContent.name}\nInput: ${JSON.stringify(toolUseContent.input)}`,
                  'Tool Use',
                  colors.magenta
                );
                console.log('');
              }
            }
          } else if (event.type === 'user') {
            const userEvent = event as UserResponse;
            for (const content of userEvent.message.content) {
              if (content.type === 'tool_result') {
                const toolResultContent = content as ToolResultContent;
                const tool_result_dict = {
                  tool_use_id: toolResultContent.tool_use_id,
                  content: toolResultContent.content,
                  is_error: toolResultContent.is_error
                };
                tool_results.push(tool_result_dict);

                const result_color = toolResultContent.is_error ? colors.red : colors.green;
                const result_icon = toolResultContent.is_error ? '❌' : '✅';

                console.log('\n');
                printPanel(
                  `${result_icon} Tool Result\n${toolResultContent.content}`,
                  'Tool Response',
                  result_color
                );
                console.log('');
              }
            }
          } else if (event.type === 'system') {
            const systemEvent = event as SystemInit;
            session_id = systemEvent.session_id;
          } else if (event.type === 'result') {
            const resultEvent = event as ResultSummary;
            cost = resultEvent.total_cost_usd;
            if (resultEvent.usage) {
              tokens = resultEvent.usage.input_tokens + resultEvent.usage.output_tokens;
            }
            success = !resultEvent.is_error;
          }
        }
      }

      const end_time = performance.now();
      const duration = (end_time - start_time) / 1000; // Convert to seconds

      // Process response
      const complete_text = response_text_parts.join('');
      const thinking_text = thinking_parts.length > 0 ? thinking_parts.join('\n') : undefined;

      // Update session metrics
      this.session_metrics.total_cost += cost;
      this.session_metrics.total_tokens += tokens;
      this.session_metrics.requests_count += 1;
      this.response_times.push(duration);
      this.session_metrics.average_response_time = 
        this.response_times.reduce((a, b) => a + b, 0) / this.response_times.length;

      // Create response object
      const simple_response = new SimpleResponseImpl({
        text: complete_text,
        cost: cost,
        tokens_used: tokens,
        model: model,
        success: success,
        session_id: session_id,
        duration_seconds: duration,
        thinking: thinking_text,
        tool_uses: tool_uses,
        tool_results: tool_results
      });

      // Show metrics if enabled
      if (this.config.show_metrics) {
        console.log('\n');
        printRule('Response Complete', colors.green);

        const metrics_text = 
          `✅ Response completed in ${duration.toFixed(1)}s\n` +
          `💰 Cost: $${cost.toFixed(4)} | 🎯 Tokens: ${tokens} | 🤖 Model: ${model}\n\n` +
          `📊 Session Stats:\n` +
          `💵 Total Cost: $${this.session_metrics.total_cost.toFixed(4)}\n` +
          `🎯 Total Tokens: ${this.session_metrics.total_tokens}\n` +
          `📨 Requests: ${this.session_metrics.requests_count}\n` +
          `⏱️  Avg Response Time: ${this.session_metrics.average_response_time.toFixed(1)}s`;

        printPanel(metrics_text, '📊 Metrics & Performance', colors.blue);
        console.log('');
      }

      return simple_response;

    } catch (error) {
      const end_time = performance.now();
      const duration = (end_time - start_time) / 1000;

      console.log('\n');
      printPanel(
        `❌ Something went wrong: ${error}\n\n🔄 Don't worry, this happens sometimes. Try again!`,
        'Error',
        colors.red
      );
      console.log('');

      return new SimpleResponseImpl({
        text: `Error: ${error}`,
        cost: 0.0,
        tokens_used: 0,
        model: "unknown",
        success: false,
        duration_seconds: duration,
        tool_uses: [],
        tool_results: []
      });
    }
  }

  /**
   * Have a multi-turn conversation with Claude
   */
  async chat(prompts: string[]): Promise<SimpleResponse[]> {
    const responses: SimpleResponse[] = [];
    console.log('\n');
    printPanel(
      `🗣️ Starting conversation with ${prompts.length} messages...`,
      '',
      colors.cyan
    );
    console.log('');

    for (let i = 0; i < prompts.length; i++) {
      const prompt = prompts[i];
      console.log('\n');
      printRule(`💬 Message ${i + 1}/${prompts.length}`, colors.magenta);
      
      const response = await this.ask(prompt);
      responses.push(response);

      if (!response.success) {
        printColored(`❌ Stopping conversation due to error in message ${i + 1}`, colors.red);
        break;
      }
    }

    console.log('\n');
    printPanel(
      `🎉 Conversation completed! ${responses.length} responses received.`,
      '',
      colors.green
    );
    console.log('');
    return responses;
  }

  /**
   * Get current session metrics
   */
  get_metrics(): SimpleMetrics {
    return this.session_metrics;
  }

  /**
   * Reset session metrics to zero
   */
  reset_metrics(): void {
    this.session_metrics = new SimpleMetricsImpl({
      total_cost: 0.0,
      total_tokens: 0,
      requests_count: 0,
      average_response_time: 0.0
    });
    this.response_times = [];
    printColored('📊 Session metrics reset!', colors.green);
  }

  /**
   * Change the Claude model
   */
  change_model(model: string): void {
    this.config.model = model;
    const fallback = model !== "haiku" ? "haiku" : "sonnet";
    this.claude_api.set_model(model, fallback);
    printColored(`🤖 Switched to model: ${model} (fallback: ${fallback})`, colors.cyan);
  }

  /**
   * Enable or disable showing Claude's thinking process
   */
  enable_thinking(enabled: boolean = true): void {
    this.config.show_thinking = enabled;
    const status = enabled ? "enabled" : "disabled";
    printColored(`🧠 Thinking display ${status}`, colors.yellow);
  }

  /**
   * Enable or disable metrics display
   */
  enable_metrics(enabled: boolean = true): void {
    this.config.show_metrics = enabled;
    const status = enabled ? "enabled" : "disabled";
    printColored(`📊 Metrics display ${status}`, colors.blue);
  }

  /**
   * Set or clear the system prompt that gets appended to all requests
   */
  set_system_prompt(system_prompt?: string): void {
    this.config.system_prompt = system_prompt;
    this.claude_api.set_system_prompt(system_prompt || "");

    if (system_prompt) {
      printColored(`📝 System prompt set: ${system_prompt}`, colors.green);
    } else {
      printColored('📝 System prompt cleared', colors.green);
    }
  }

  /**
   * Show help information
   */
  help(): void {
    const help_text = `
🤖 Simple Claude API Help

Basic Usage:
  const claude = new SimpleClaudeAPI();
  const response = await claude.ask("Your question here");

Methods:
  .ask(prompt, callback?)         - Ask a question (with streaming)
  .chat([prompt1, prompt2, ...])  - Multi-turn conversation
  .get_metrics()                  - View session statistics
  .reset_metrics()                - Reset statistics
  .change_model(model)            - Switch models
  .enable_thinking(true/false)    - Show/hide thinking
  .enable_metrics(true/false)     - Show/hide metrics
  .set_system_prompt(prompt)      - Set system prompt for all requests
  .help()                         - Show this help

Examples:
  // Basic usage
  await claude.ask("What is Python?");
  
  // With custom callback
  const my_callback = (text: string) => {
    console.log(\`Claude: \${text}\`);
  };
  await claude.ask("Tell me a joke", my_callback);
  
  // Multiple questions
  await claude.chat(["Hi!", "What's the weather like?", "Thanks!"]);
  
  // With system prompt
  claude.set_system_prompt("You are a helpful Python tutor.");
  await claude.ask("How do I create a list?");
  
  // Initialize with system prompt
  import { SimpleClaudeAPI, SimpleConfigImpl } from './simple_api';
  const config = new SimpleConfigImpl({
    system_prompt: "You are a creative writer."
  });
  const claude = new SimpleClaudeAPI(config);

Models available: 'sonnet' (default), 'opus', 'haiku'
    `;

    printPanel(help_text, '🤖 Simple Claude API Help', colors.blue);
  }
}

// Export the config class for external use
export { SimpleConfigImpl };

// Convenience function for one-off questions
export async function ask_claude_simple(
  prompt: string,
  model: string = "sonnet",
  show_thinking: boolean = false,
  system_prompt?: string
): Promise<SimpleResponse> {
  const config = new SimpleConfigImpl({
    model: model,
    show_thinking: show_thinking,
    system_prompt: system_prompt
  });
  const claude = new SimpleClaudeAPI(config);
  return claude.ask(prompt);
}

// Even simpler function that just prints the answer
export async function claude_say(prompt: string): Promise<void> {
  await ask_claude_simple(prompt);
}